docs/doxygen/man/man3/CreateAccount.3
.TH "CreateAccount" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CreateAccount \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCreateAccount\fP (uint16_t src_bank, uint32_t src_user, uint32_t msg_id, uint16_t dst_bank, uint32_t time)"
.br
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_USR as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eModifying as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetAdditionalData\fP () override"
.br
.RI "\fIGet additional data\&. Used only on server side to add new created account data\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual int \fBgetAdditionalDataSize\fP () override"
.br
.RI "\fIGet additional data size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "virtual uint32_t \fBgetDestBankId\fP ()"
.br
.RI "\fIGet destination bank id\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet user message id\&. \fP"
.ti -1c
.RI "virtual void \fBsetAdditionalData\fP (char *data)"
.br
.RI "\fIPut new created account data\&. \fP"
.ti -1c
.RI "virtual void \fBsetNewUser\fP (uint32_t userId, uint8_t userPkey[])"
.br
.RI "\fISetting new account data\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetPublicKey\fP ()"
.br
.RI "\fIGet public key from additional data\&. \fP"
.ti -1c
.RI "virtual void \fBsetPublicKey\fP (uint8_t user_pkey[SHA256_DIGEST_LENGTH])"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Definition at line 9 of file createaccount\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool CreateAccount::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 65 of file createaccount\&.cpp\&.
.SS "bool CreateAccount::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 110 of file createaccount\&.cpp\&.
.SS "void CreateAccount::setNewUser (uint32_t userId, uint8_t userPkey[])\fC [virtual]\fP"

.PP
Setting new account data\&. 
.PP
\fBParameters:\fP
.RS 4
\fIuserId\fP - new account user id 
.br
\fIuserPkey\fP - new account public key 
.RE
.PP

.PP
Definition at line 171 of file createaccount\&.cpp\&.
.SS "void CreateAccount::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 61 of file createaccount\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/createaccount.cpp
#include "createaccount.h"
#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/hash.h"
#include "helper/json.h"

CreateAccount::CreateAccount()
    : m_data{} {
    m_newAccount.user_id = 0;
    memset(m_newAccount.user_pkey, 0, sizeof(m_newAccount.user_pkey));
    m_responseError = ErrorCodes::Code::eNone;
}

CreateAccount::CreateAccount(uint16_t src_bank, uint32_t src_user, uint32_t msg_id, uint16_t dst_bank, uint32_t time)
    : m_data(src_bank, src_user, msg_id, time, dst_bank) {
    m_newAccount.user_id = 0;
    memset(m_newAccount.user_pkey, 0, sizeof(m_newAccount.user_pkey));
    m_responseError = ErrorCodes::Code::eNone;
}

unsigned char* CreateAccount::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* CreateAccount::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_response);
}

void CreateAccount::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void CreateAccount::setResponse(char* response) {
    m_response = *reinterpret_cast<decltype(m_response)*>(response);
}

int CreateAccount::getDataSize() {
    return sizeof(m_data.info);
}

int CreateAccount::getResponseSize() {
    return sizeof(m_response);
}

unsigned char* CreateAccount::getSignature() {
    return m_data.sign;
}

int CreateAccount::getSignatureSize() {
    return sizeof(m_data.sign);
}

int CreateAccount::getType() {
    return TXSTYPE_USR;
}

CommandType CreateAccount::getCommandType() {
    return CommandType::eModifying;
}

void CreateAccount::sign(const uint8_t* hash, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign2(hash,SHA256_DIGEST_LENGTH , getData(), getDataSize(), sk, pk, getSignature());
}

bool CreateAccount::checkSignature(const uint8_t* hash, const uint8_t* pk) {
    return (ed25519_sign_open2(hash, SHA256_DIGEST_LENGTH,getData(),getDataSize(),pk,getSignature()) == 0);
}

void CreateAccount::saveResponse(settings& sts) {
    if (!sts.without_secret && !std::equal(sts.pk, sts.pk + SHA256_DIGEST_LENGTH, m_response.usera.pkey)) {
        m_responseError = ErrorCodes::Code::ePkeyDiffers;
    }

    std::array<uint8_t, SHA256_DIGEST_LENGTH> hashout;
    Helper::create256signhash(getSignature(), getSignatureSize(), sts.ha, hashout);
    if (!sts.signature_provided && !std::equal(hashout.begin(), hashout.end(), m_response.usera.hash)) {
        m_responseError = ErrorCodes::Code::eHashMismatch;
    }

    sts.msid = m_response.usera.msid;
    std::copy(m_response.usera.hash, m_response.usera.hash + SHA256_DIGEST_LENGTH, sts.ha.data());
}

uint32_t CreateAccount::getUserId() {
    return m_data.info.src_user;
}

uint32_t CreateAccount::getBankId() {
    return m_data.info.src_node;
}

uint32_t CreateAccount::getTime() {
    return m_data.info.ttime;
}

int64_t CreateAccount::getFee() {
    int64_t fee;
    if (this->getBankId() == this->getDestBankId()) {
        fee = TXS_USR_FEE;
    } else {
        fee = TXS_USR_FEE + TXS_RUS_FEE;
    }
    return fee;
}

int64_t CreateAccount::getDeduct() {
    return USER_MIN_MASS;
}

bool CreateAccount::send(INetworkClient& netClient) {
    if(!netClient.sendData(getData(), getDataSize())) {
        ELOG("CreateAccount sending data error\n");
        return false;
    }

    if(!netClient.sendData(getAdditionalData(), getAdditionalDataSize())) {
        ELOG("CreateAccount sending additional data error\n");
        return false;
    }

    if(!netClient.sendData(getSignature(), getSignatureSize())) {
        ELOG("CreateAccount sending signature error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
       ELOG("CreateAccount reading error\n");
       return false;
    }

    if (m_responseError) {
        return true;
    }

    if(!netClient.readData(getResponse(), getResponseSize())) {
        ELOG("CreateAccount ERROR reading global info\n");
        return false;
    }

    return true;
}

unsigned char* CreateAccount::getAdditionalData() {
    return reinterpret_cast<unsigned char*>(&m_newAccount);
}

void CreateAccount::setAdditionalData(char* data) {
    m_newAccount = *reinterpret_cast<decltype(m_newAccount)*>(data);
}

int CreateAccount::getAdditionalDataSize() {
    return sizeof(m_newAccount);
}

unsigned char* CreateAccount::getPublicKey() {
    return reinterpret_cast<unsigned char*>(&m_newAccount.user_pkey);
}

void CreateAccount::setPublicKey(uint8_t user_pkey[SHA256_DIGEST_LENGTH]) {
    memcpy(m_newAccount.user_pkey, user_pkey, SHA256_DIGEST_LENGTH);
}

uint32_t CreateAccount::getDestBankId() {
    return m_data.info.dst_node;
}

uint32_t CreateAccount::getUserMessageId() {
    return m_data.info.msg_id;
}

void CreateAccount::setNewUser(uint32_t userId, uint8_t userPkey[SHA256_DIGEST_LENGTH]) {
    m_newAccount = NewAccountData(userId, userPkey);
}

std::string CreateAccount::toString(bool /*pretty*/) {
    return "";
}

void CreateAccount::toJson(boost::property_tree::ptree& ptree) {
    if (!m_responseError) {
        Helper::print_user(m_response.usera, ptree, true, this->getBankId(), this->getUserId());
        Helper::print_msgid_info(ptree, m_data.info.src_node, m_response.msid, m_response.mpos);
        // only for account created in the same node
        if (m_response.new_user) {
            char nAccountAddress[19]="";
            uint16_t suffix=Helper::crc_acnt(getBankId(), m_response.new_user);
            sprintf(nAccountAddress, "%04X-%08X-%04X", getBankId(), m_response.new_user, suffix);
            ptree.put("new_account.address", nAccountAddress);
            ptree.put("new_account.node", this->getBankId());
            ptree.put("new_account.id", m_response.new_user);
        }
    } else {
        if (m_responseError == ErrorCodes::Code::ePkeyDiffers) {
            std::stringstream tx_user_hashin{};
            Helper::ed25519_key2text(tx_user_hashin, m_response.usera.pkey, SHA256_DIGEST_LENGTH);
            ptree.put("tx.account_public_key_new", tx_user_hashin.str());
        }
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    }
}

void CreateAccount::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.src_node);
    ptree.put(TAG::SRC_USER, m_data.info.src_user);
    ptree.put(TAG::MSGID, m_data.info.msg_id);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::DST_NODE, m_data.info.dst_node);
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string CreateAccount::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"create_account\",[\"node\":<node id>]}" << "\n";
    ss << "Example: " << "{\"run\":\"create_account\",\"node\":\"2\"}" << "\n";
    return ss.str();
}
